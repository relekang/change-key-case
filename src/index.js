// @flow
import changeCase from "change-case";
import curry from "curry";

export type CaseType =
  | "camel"
  | "constant"
  | "dot"
  | "header"
  | "lower"
  | "lcFirst"
  | "no"
  | "param"
  | "pascal"
  | "path"
  | "sentence"
  | "snake"
  | "swap"
  | "title"
  | "upper"
  | "ucFirst";

type ChangeKeyCase = ((caseType: CaseType) => (o: Object) => Object) &
  ((caseType: CaseType, o: Object) => Object);

export const changeKeyCase: ChangeKeyCase = curry(function changeKeyCase(
  caseType: CaseType,
  o: Object
): Object {
  return Object.keys(o).reduce(
    (lastValue, key) => ({
      ...lastValue,
      [changeCase[caseType + "Case"](key)]: o[key]
    }),
    {}
  );
});
