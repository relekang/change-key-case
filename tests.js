// @flow
import { changeKeyCase } from "./src";

test("changeKeyCase should change the case of every key in the object with partial application", () => {
  expect(
    changeKeyCase("sentence")({
      accountType: "standard",
      amount: 200,
      transactionTypeCode: 1
    })
  ).toEqual({
    "Account type": "standard",
    Amount: 200,
    "Transaction type code": 1
  });
});

test("changeKeyCase should change the case of every key in the object", () => {
  expect(
    changeKeyCase("sentence", {
      accountType: "standard",
      amount: 200,
      transactionTypeCode: 1
    })
  ).toEqual({
    "Account type": "standard",
    Amount: 200,
    "Transaction type code": 1
  });
});

const testObject = {
  transactionTypeCode: 1,
  account_type: "standard"
};

[
  "camel",
  "constant",
  "dot",
  "header",
  "lower",
  "no",
  "param",
  "pascal",
  "path",
  "sentence",
  "snake",
  "swap",
  "title",
  "upper"
].forEach(type => {
  test("changeKeyCase(" + type + ")(TestObject)", () => {
    expect(changeKeyCase(type)(testObject)).toMatchSnapshot();
  });
});
