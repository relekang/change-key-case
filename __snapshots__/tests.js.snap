// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`changeKeyCase(camel)(TestObject) 1`] = `
Object {
  "accountType": "standard",
  "transactionTypeCode": 1,
}
`;

exports[`changeKeyCase(constant)(TestObject) 1`] = `
Object {
  "ACCOUNT_TYPE": "standard",
  "TRANSACTION_TYPE_CODE": 1,
}
`;

exports[`changeKeyCase(dot)(TestObject) 1`] = `
Object {
  "account.type": "standard",
  "transaction.type.code": 1,
}
`;

exports[`changeKeyCase(header)(TestObject) 1`] = `
Object {
  "Account-Type": "standard",
  "Transaction-Type-Code": 1,
}
`;

exports[`changeKeyCase(lower)(TestObject) 1`] = `
Object {
  "account_type": "standard",
  "transactiontypecode": 1,
}
`;

exports[`changeKeyCase(no)(TestObject) 1`] = `
Object {
  "account type": "standard",
  "transaction type code": 1,
}
`;

exports[`changeKeyCase(param)(TestObject) 1`] = `
Object {
  "account-type": "standard",
  "transaction-type-code": 1,
}
`;

exports[`changeKeyCase(pascal)(TestObject) 1`] = `
Object {
  "AccountType": "standard",
  "TransactionTypeCode": 1,
}
`;

exports[`changeKeyCase(path)(TestObject) 1`] = `
Object {
  "account/type": "standard",
  "transaction/type/code": 1,
}
`;

exports[`changeKeyCase(sentence)(TestObject) 1`] = `
Object {
  "Account type": "standard",
  "Transaction type code": 1,
}
`;

exports[`changeKeyCase(snake)(TestObject) 1`] = `
Object {
  "account_type": "standard",
  "transaction_type_code": 1,
}
`;

exports[`changeKeyCase(swap)(TestObject) 1`] = `
Object {
  "ACCOUNT_TYPE": "standard",
  "TRANSACTIONtYPEcODE": 1,
}
`;

exports[`changeKeyCase(title)(TestObject) 1`] = `
Object {
  "Account Type": "standard",
  "Transaction Type Code": 1,
}
`;

exports[`changeKeyCase(upper)(TestObject) 1`] = `
Object {
  "ACCOUNT_TYPE": "standard",
  "TRANSACTIONTYPECODE": 1,
}
`;
